src/mon/PGMap.cc

```diff
@@ -2548,6 +2550,629 @@ namespace {
+      }
+      if (!(i.second.state & PG_STATE_CLEAN) &&
+	  i.second.last_clean < cutoff) {
+	if (!unclean) {
+	  unclean = &checks->add("PG_STUCK_UNCLEAN", HEALTH_ERR, "");
+	}
+	note_stuck_detail(i.first, i.second, "unclean", i.second.last_clean,
+			  now, max, &unclean->detail);
+	++num_unclean;
+      }
+      if ((i.second.state & PG_STATE_DEGRADED) &&
+	  i.second.last_undegraded < cutoff) {
+	if (!degraded) {
+	  degraded = &checks->add("PG_STUCK_DEGRADED", HEALTH_ERR, "");
+	}
+	note_stuck_detail(i.first, i.second, "degraded",
+			  i.second.last_undegraded, now, max, &degraded->detail);
+	++num_degraded;
+      }
+      if ((i.second.state & PG_STATE_UNDERSIZED) &&
+	  i.second.last_fullsized < cutoff) {
+	if (!undersized) {
+	  undersized = &checks->add("PG_STUCK_UNDERSIZED", HEALTH_ERR, "");
+	}
+	note_stuck_detail(i.first, i.second, "undersized",
+			  i.second.last_fullsized, now, max,
+			  &undersized->detail);
+	++num_undersized;
+      }
+      if ((i.second.state & PG_STATE_STALE) &&
+	  i.second.last_unstale < cutoff) {
+	if (!stale) {
+	  stale = &checks->add("PG_STUCK_STALE", HEALTH_ERR, "");
+	}
+	note_stuck_detail(i.first, i.second, "stale", i.second.last_unstale,
+			  now, max, &stale->detail);
+	++num_stale;
+      }
+    }
+    if (inactive) {
+      note_stuck_summary("inactive", num_inactive, max, inactive);
+    }
+    if (unclean) {
+      note_stuck_summary("unclean", num_unclean, max, unclean);
+    }
+    if (degraded) {
+      note_stuck_summary("degraded", num_degraded, max, degraded);
+    }
+    if (undersized) {
+      note_stuck_summary("undersized", num_undersized, max, undersized);
+    }
+    if (stale) {
+      note_stuck_summary("stale", num_stale, max, stale);
+    }
+  }
+
+  // OSD_SKEWED_USAGE
+  if (cct->_conf->mon_warn_osd_usage_min_max_delta) {
+    int max_osd = -1, min_osd = -1;
+    float max_osd_usage = 0.0, min_osd_usage = 1.0;
+    for (auto p = osd_stat.begin(); p != osd_stat.end(); ++p) {
+      // kb should never be 0, but avoid divide by zero in case of corruption
+      if (p->second.kb <= 0)
+        continue;
+      float usage = ((float)p->second.kb_used) / ((float)p->second.kb);
+      if (usage > max_osd_usage) {
+        max_osd_usage = usage;
+	max_osd = p->first;
+      }
+      if (usage < min_osd_usage) {
+        min_osd_usage = usage;
+	min_osd = p->first;
+      }
+    }
+    float diff = max_osd_usage - min_osd_usage;
+    if (diff > cct->_conf->mon_warn_osd_usage_min_max_delta) {
+      auto& d = checks->add("OSD_SKEWED_USAGE", HEALTH_WARN,
+			    "skewed osd utilization");
+      ostringstream ss;
+      ss << "difference between min (osd." << min_osd << " at "
+	 << roundf(min_osd_usage*1000.0)/100.0
+	 << "%) and max (osd." << max_osd << " at "
+	 << roundf(max_osd_usage*1000.0)/100.0
+	 << "%) osd usage " << roundf(diff*1000.0)/100.0 << "% > "
+	 << roundf(cct->_conf->mon_warn_osd_usage_min_max_delta*1000.0)/100.0
+	 << " (mon_warn_osd_usage_min_max_delta)";
```

Comment:
This should probably include a link to something explaining PG and OSD balancing.